Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,342 @@
+there exists a group homomorphims `H → G` which is a closed embedding compatible with the actions,
+then `H` also acts properly on `X`. -/
+@[to_additive "If two groups `H` and `G` act on a topological space `X` such that `G` acts properly
+and there exists a group homomorphims `H → G` which is a closed embedding compatible with the
+actions, then `H` also acts properly on `X`."]
+lemma properSMul_of_closed_embedding {H : Type*} [Group H] [MulAction H X] [TopologicalSpace H]
+    [ProperSMul G X] (f : H →* G) (f_clemb : ClosedEmbedding f)
+    (f_compat : ∀ (h : H) (x : X), f h • x = h • x) : ProperSMul H X where
+  isProperMap_smul_pair' := by
+    have := isProperMap_of_closedEmbedding f_clemb
+    have : IsProperMap (Prod.map f (fun x : X ↦ x)) := IsProperMap.prod_map this isProperMap_id
+    have : (fun hx : H × X ↦ (hx.1 • hx.2, hx.2)) = (fun hx ↦ (f hx.1 • hx.2, hx.2)) := by
+      simp [f_compat]
+    rw [this]
+    change IsProperMap ((fun gx : G × X ↦ (gx.1 • gx.2, gx.2)) ∘ (Prod.map f (fun x ↦ x)))
+    apply IsProperMap.comp
+    assumption
+    exact isProperMap_smul_pair G X
+
+/-- If `H` is a closed subgroup of `G` and `G` acts properly on X then so does `H`. -/
+@[to_additive "If `H` is a closed subgroup of `G` and `G` acts properly on X then so does `H`."]
+instance {H : Subgroup G} [ProperSMul G X] [H_closed : IsClosed (H : Set G)] : ProperSMul H X where
+  isProperMap_smul_pair' := by
+    have : IsProperMap (fun hx : H × X ↦ ((hx.1, hx.2) : G × X)) := by
+      change IsProperMap (Prod.map ((↑) : H → G) (fun x ↦ x))
+      exact IsProperMap.prod_map (isProperMap_subtype_val_of_closed H_closed) isProperMap_id
+    have : IsProperMap (fun hx : H × X ↦ (hx.1 • hx.2, hx.2)) := by
+      change IsProperMap ((fun gx ↦ (gx.1 • gx.2, gx.2)) ∘
+        (fun hx : H × X ↦ ((hx.1, hx.2) : G × X)))
+      exact this.comp (isProperMap_smul_pair G X)
+    assumption
+
+/-
+Some suggestions of things to prove,
+taken from Kapovich
+-/
+
+/-
+If `X` is T2 and first countable,
+then the naive definition
+of proper action is equivalent to the good definition
+(Kapovich Th 11)
+-/
+theorem naiveProper_iff_ProperSMul_T2_FirstCountable
+    [T2Space X] [FirstCountableTopology X] :
+    ProperlyDiscontinuousSMul G X ↔ ProperSMul G X
+    := by sorry
+
+lemma tendsTo_comp_continuous
+    {lx: Filter X} {f : X → Y} {g : Y → Z} {y : Y}
+    (H : Tendsto f lx (𝓝 y)) (hg: Continuous g) :
+    Tendsto (g ∘ f) lx (𝓝 (g y)) := by
+  apply Filter.Tendsto.comp _ H
+  exact hg.tendsto y
+
+/-
+If `X` and `Y` are T2 and first countable,
+then the naive definition
+of proper map is equivalent to the good definition
+-/
+theorem properMap_of_naiveProper_T2_FirstCountable
+    [T2Space X] [FirstCountableTopology X]
+    [T2Space Y] [FirstCountableTopology Y]
+    (f : X → Y) (hcont: Continuous f):
+    (∀ (K : Set Y), (IsCompact K → IsCompact (f ⁻¹' K)))
+    → IsProperMap f
+    := by
```

Comment:
In general, we tend to write all hypotheses before the colon, to avoid having to introduce them. This would mean something like:
```suggestion
    (f : X → Y) (hcont: Continuous f)
    (h : ∀ (K : Set Y), (IsCompact K → IsCompact (f ⁻¹' K))) : IsProperMap f := by
```

Also, note that the linter doesn't want to have line breaks just before `:= by`.